src/client/js/client.js

```diff
@@ -1409,67 +1303,211 @@ define([
+                if (state.inLoading) {
+                    state.ongoingLoadPatternsCounter += 1;
+                    loadPatternThrottled(state.core,
+                        keys[i], patterns[keys[i]], state.loadNodes, loadingPatternFinished);
                 } else {
-                    //we should update the patterns, but that is all
-                    state.users[guid].PATTERNS = JSON.parse(JSON.stringify(patterns));
+                    //we should save the patterns to a pending directory
+                    state.pendingTerritoryUpdatePatterns[updateRequestId] = patterns;
                 }
+                loadPatternThrottled(state.core, keys[i], patterns[keys[i]], state.nodes, patternLoaded);
             }
+
         };
 
+        function refreshMetaNodes(oldSource, newSource) {
+            var pathsToRemove = [],
+                i,
+                oldPaths = Object.keys(oldSource),
+                newPaths = Object.keys(newSource);
+
+            for (i = 0; i < oldPaths.length; i += 1) {
+                if (newPaths.indexOf(oldPaths[i]) === -1) {
+                    pathsToRemove.push(oldPaths[i]);
+                }
+            }
+
+            for (i = 0; i < newPaths.length; i += 1) {
+                state.metaNodes[newPaths[i]] = newSource[newPaths[i]].node;
+            }
+
+            for (i = 0; i < pathsToRemove.length; i += 1) {
+                delete state.metaNodes[pathsToRemove[i]];
+            }
+        }
+
+        function switchStates() {
+            //it is safe now to move the loadNodes into nodes,
+            // refresh the metaNodes and generate events - all in a synchronous manner!!!
+            var modifiedPaths,
+                i;
+
+            refreshMetaNodes(state.nodes, state.loadNodes);
+
+            modifiedPaths = getModifiedNodes(state.loadNodes);
+            state.nodes = state.loadNodes;
+            state.loadNodes = {};
+
+            state.inLoading = false;
+            state.rootHash = state.loading.rootHash;
+            state.loading.rootHash = null;
+            state.commitHash = state.loading.commitHash;
+            state.loading.commitHash = null;
+
+            for (i in state.users) {
+                if (state.users.hasOwnProperty(i)) {
+                    userEvents(i, modifiedPaths);
+                }
+            }
+
+            if (state.loadingStatus) {
+                state.loading.next(state.loadingStatus);
+            } else {
+                addOnFunctions.updateRunningAddOns(state.nodes[ROOT_PATH], state.loading.next);
+            }
+        }
+
+        function loading(newRootHash, newCommitHash, callback) {
+            var i, j,
+                userIds,
+                patternPaths,
+                patternsToLoad = [];
+
+            if (state.ongoingLoadPatternsCounter !== 0) {
+                logger.error('at the start of loading the counter should be zero!!! [' + state.ongoingTerritoryUpdateCounter + ']');

```

Comment:
If we're f-ed here, why not throw an Exception with the same error message. Some stack-trace couldn't heart..